### PR TITLE
NIFI-9691: Added ForkEnrichment, JoinEnrichment processors

### PR DESCRIPTION
**Pull request by markap14**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by gresockj — CHANGES_REQUESTED** [review]
Great addition, @markap14!  This is a really useful set of processors that seems to have a wide application.

I have a few minor typo and phrasing suggestions, and ran into an issue where the timeout did not seem to work when I set it to 30 seconds.  Otherwise, the code looks good, and I ran all the input samples from the JoinProcessor Additional Details documentation and it generated the expected results.  Nice work!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/JoinEnrichment.java [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+    static final PropertyDescriptor RECORD_WRITER = new PropertyDescriptor.Builder()
+        .name("Record Writer")
+        .displayName("Record Writer")
+        .description("The Record Writer to use for writing the results. If the Record Writer is configured to inherit the schema from the Record, the schema that it will inherit will be the result " +
+            "of merging both the 'original' record schema and the 'enrichment' record schema.")
+        .required(true)
+        .identifiesControllerService(RecordSetWriterFactory.class)
+        .build();
+
+    // TODO: Include in docs: What to do if we need something more complex like a script: Use Wrapper strategy to correlate the records and then use a ScriptedTransformRecord.
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/JoinEnrichment.java:138]: Yup :) Thanks.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/enrichment/IndexCorrelatedJoinStrategy.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                private boolean usedFirstRecords = false;
+
+                @Override
+                public RecordSchema getSchema() {
+                    return resultSchema;
+                }
+
+                @Override
+                public Record next() throws IOException {
+                    if (!usedFirstRecords) {
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/enrichment/IndexCorrelatedJoinStrategy.java:80]: So, in order to derive the schema for the Record Writer, we need to get the schema from both the 'original' and 'enrichment' readers. We need to then combine the records together to derive the output schema. In the case of `InsertRecordFieldsJoinStrategy`. The only way to do this is to obtain a record from each reader and then combine them and call `incorporateInactiveFields` on the resultant Record.

So we do that to derive the output schema for the RecordWriter. But now we've already consumed the first record. So the first time through, we need to return that first Record. The rest of the times through, we just return the next Records.

**Comment by markap14** [comment]
Thanks for the thorough review @gresockj ! I agree with all points. Updated everything, I believe. I also managed to reproduce the issue you saw with the timeout not working. Only happens when the number of FlowFiles Queued = Max Number of Bins. That's been addressed.

**Comment by gresockj** [comment]
> Thanks for the thorough review @gresockj ! I agree with all points. Updated everything, I believe. I also managed to reproduce the issue you saw with the timeout not working. Only happens when the number of FlowFiles Queued = Max Number of Bins. That's been addressed.

Thanks for the updates @markap14!  This did resolve the timeout issue with 1 bin.  However, I now see some MergeContent-related test failures that I wonder if are related to this change.

**Review by gresockj — APPROVED** [review]
Thanks for fixing those tests.  Will merge!

**Pull request closed without merging**